File: public/js/game/model/room.js
function Room(data){
    if(!data)
        data = {}
    this.roomId = data.roomId
    this.users = null;
    this.setUsers = function(usersData){
        this.users = []
        var user = Model.user;
        for(var i in usersData){
            var userData = usersData[i];
            user = GameUser.instantiate(userData)
            user.room = this;
            this.users.push(user);
        }
    }
    this.setUsers(data.users || []);
    this.addUser = function(gameUser){
        this.users.push(gameUser);
        gameUser.room = this;
    }
    this.removeUser = function(gameUser){
        var users = this.users;
        var deleted = false;
        for(var i in users){
            var user2 = users[i];
            if(user2.uid == gameUser.uid){
                users.splice(i, 1);
                user2.room = null;
                deleted = user2;
                break;
            }
        }
        return deleted
    }
    this.containsUser = function(gameUser){
        var users = this.users;
        for(var i in users)
            if(users[i].uid == gameUser.uid)
                return users[i];
        return false
    }
    this.owner = function(){
        return this.users[0];
    }
    this.findUserById = function(uid){
        var users = this.users;
        for(var i in users)
            if(users[i].uid == uid)
                return users[i];
        return null;
    }
}

File: src/main/java/com/hellsepontus/model/Room.java
package com.hellsepontus.model;

import com.corundumstudio.socketio.SocketIOClient;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Room implements IJson {
    private static int roomIdCounter = 0;
    private int roomId;
    private ArrayList<GameUser> users;
    private IRoomDestroyer destroyer;
    
    private Room(IRoomDestroyer destroyer){
        users = new ArrayList();
        this.destroyer = destroyer;
    }

    protected static Room instantiate(IRoomDestroyer destroyer){
        return instantiate(++roomIdCounter, destroyer);
    }
    
    protected static Room instantiate(int id, IRoomDestroyer destroyer){
        Room room = new Room(destroyer);
        room.roomId = id;
        return room;
    }

    public GameUser findUserByClient(SocketIOClient client) {
        for(GameUser user : users)
            if(user.getClient().equals(client))
                return user;
        return null;
    }
    
    public void addUser(GameUser user){
        users.add(user);
        user.setRoom(this);
    }
    
    public void removeUser(GameUser user){
        if(user.isOwner()){
            destroyer.destroyRoom(this.roomId);
            destroyer = null;
        }else {
            users.remove(user);
            user.setRoom(null);
        }
    }
    
    public List<GameUser> users(){
        return users;
    }
    
    public boolean containsUser(int uid) {
        for(GameUser user : users)
            if(user.uid == uid)
                return true;
        return false;
    }

    public Map<String, Object> toData() {
        Map<String, Object> result = new HashMap<String, Object>();
        result.put("roomId", this.roomId);
        List<Object> users = new ArrayList<Object>();
        for(GameUser u : this.users)
            users.add(u.toData());
        result.put("users", users);
        return result;
    }

    public void fromData(Map<String, Object> data) {
        throw new NoSuchMethodError();
    }

    public GameUser owner() {
        return users.get(0);
    }
    
    public interface IRoomDestroyer{
        void destroyRoom(int roomId);
    }
}
